flows/runs/environment.go

```diff
@@ -25,22 +25,45 @@ func newRunEnvironment(base envs.Environment, run *flowRun) *runEnvironment {
 func (e *runEnvironment) Timezone() *time.Location {
 	contact := e.run.Contact()
 
-	// if run has a contact with a timezone, that overrides the enviroment's timezone
+	// if we have a contact and they have a timezone that overrides the base enviroment's timezone
 	if contact != nil && contact.Timezone() != nil {
 		return contact.Timezone()
 	}
-	return e.run.Session().Environment().Timezone()
+	return e.Environment.Timezone()
+}
+
+func (e *runEnvironment) DefaultLanguage() envs.Language {
+	contact := e.run.Contact()
+
+	// if we have a contact and they have a language and it's an allowed language that overrides the base environment's languuage
+	if contact != nil && contact.Language() != envs.NilLanguage && isAllowedLanguage(e, contact.Language()) {
+		return contact.Language()
+	}
+	return e.Environment.DefaultLanguage()
 }
 
 func (e *runEnvironment) DefaultCountry() envs.Country {
 	contact := e.run.Contact()
 
-	// if run has a contact with a preferred channel with a country, that overrides the environment's country
+	// if we have a contact and they have a preferred channel with a country that overrides the base environment's country
 	if contact != nil {
 		cc := contact.Country()
 		if cc != envs.NilCountry {
 			return cc
 		}
 	}
-	return e.run.Session().Environment().DefaultCountry()
+	return e.Environment.DefaultCountry()
+}
+
+func (e *runEnvironment) DefaultLocale() envs.Locale {
```

Comment:
the other fix needed was redeclaring `DefaultLocale` otherwise with the lack of polymorphism it's the method on the base env using the base env's values